conmon/conmon.c

```diff
@@ -93,25 +93,90 @@ static GOptionEntry entries[] =
   { NULL }
 };
 
-int set_k8s_timestamp(char *buf, ssize_t buflen, const char *stream_type)
+/* "1997-03-25T13:20:42+01:00 " */
+#define TSBUFLEN 27
+
+int set_k8s_timestamp(char *buf, ssize_t buflen)
 {
 	time_t now = time(NULL);
 	struct tm *tm;
 	char off_sign = '+';
-	int off;
+	int off, len, err = -1;
+
+	if ((tm = localtime(&now)) == NULL)
+		return err;
 
-	if ((tm = localtime(&now)) == NULL) {
-		return -1;
-	}
 	off = (int) tm->tm_gmtoff;
 	if (tm->tm_gmtoff < 0) {
 		off_sign = '-';
 		off = -off;
 	}
-	snprintf(buf, buflen, "%d-%02d-%02dT%02d:%02d:%02d%c%02d:%02d %s ",
-		tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday,
-		tm->tm_hour, tm->tm_min, tm->tm_sec,
-		off_sign, off / 3600, off % 3600, stream_type);
+
+	len = snprintf(buf, buflen, "%d-%02d-%02dT%02d:%02d:%02d%c%02d:%02d ",
+		       tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday,
+		       tm->tm_hour, tm->tm_min, tm->tm_sec,
+		       off_sign, off / 3600, off % 3600);
+
+	if (len < buflen)
+		err = 0;
+	return err;
+}
+
+/*
+ * The CRI requires us to write logs with a (timestamp, stream, line) format
+ * for every newline-separated line. write_k8s_log writes said format for every
+ * line in buf, and will partially write the final line of the log if buf is
+ * not terminated by a newline.
+ *
+ * TODO: The above trick (partially writing the last line) won't work if we
+ *       ever switch to having different streams for std{out,err}. Luckily at
+ *       the moment that isn't an issue for us, but we need to fix this in the
+ *       future.
+ */
+int write_k8s_log(int fd, const char *stream_type, const char *buf, ssize_t buflen)
+{
+	char tsbuf[TSBUFLEN];
+
+	/*
+	 * Use the same timestamp for every line of the log in this buffer.
+	 * There is no practical difference in the output since write(2) is
+	 * fast.
+	 */
+	if (set_k8s_timestamp(tsbuf, TSBUFLEN))
+		/* TODO: We should handle failures much more cleanly than this. */
+		return -1;
+
+	/* XXX: How do we deal with the first line? */
+
+	while (buflen > 0) {
+		/* Find the end of the line. */
+		const char *line_end = memchr(buf, buflen, '\n');
+		if (!line_end)
+			line_end = &buf[buflen-1];
+		else
+			line_end++;
+		ptrdiff_t line_len = buf - line_end;
+
+		/* Write the (timestamp, stream, line) tuple. */
+		if (write(fd, tsbuf, TSBUFLEN-1) < 0) {
```

Comment:
Is this making sure we drop the NULL terminator which got in with snprintf? One of the original issue is that the NULL terminator (00) is causing strings matching to fail